File: ObjectiveC/MDMNotificationService/NotificationService.m
//
//  NotificationService.m
//  MDMNotificationService
//
//  Created by Jorge Silveira on 06/09/18.
//  Copyright © 2018 Rogerio. All rights reserved.
//

#import "NotificationService.h"

@interface NotificationService ()

@property (nonatomic, strong) void (^contentHandler)(UNNotificationContent *contentToDeliver);
@property (nonatomic, strong) UNMutableNotificationContent *bestAttemptContent;

@end

@implementation NotificationService

- (void)didReceiveNotificationRequest:(UNNotificationRequest *)request withContentHandler:(void (^)(UNNotificationContent * _Nonnull))contentHandler {
    
    // Debug Mode
    [MDMCore setDebugMode:YES];
    
    self.contentHandler = contentHandler;
    self.bestAttemptContent = [request.content mutableCopy];

    if ([MDMNotification isMDMNotification:self.bestAttemptContent.userInfo]) {
        [super didReceiveNotificationRequest:request withContentHandler:contentHandler];
    } else {
        // Your code here
        self.contentHandler(self.bestAttemptContent);
    }
}

- (void)serviceExtensionTimeWillExpire {
    
    // Debug Mode
    [MDMCore setDebugMode:YES];
    
    if ([MDMNotification isMDMNotification:self.bestAttemptContent.userInfo]) {
        [super serviceExtensionTimeWillExpire];
    } else {
        // Your code here
        self.contentHandler(self.bestAttemptContent);
    }
}

@end

File: Swift/MDMNotificationService/NotificationService.swift
//
//  NotificationService.swift
//  MDMNotificationService
//
//  Created by Robson Moreira on 12/12/19.
//  Copyright © 2019 Hands. All rights reserved.
//

import MDMNotification

class NotificationService: MDMNotificationService {

    var contentHandler: ((UNNotificationContent) -> Void)?
    var bestAttemptContent: UNMutableNotificationContent?

    override func didReceive(_ request: UNNotificationRequest, withContentHandler contentHandler: @escaping (UNNotificationContent) -> Void) {
        self.contentHandler = contentHandler
        self.bestAttemptContent = (request.content.mutableCopy() as? UNMutableNotificationContent)

        if let bestAttemptContent = self.bestAttemptContent {
            if MDMNotification.isMDMNotification(bestAttemptContent.userInfo) {
                super.didReceive(request, withContentHandler: contentHandler)
            } else {
                // Your code here
                contentHandler(bestAttemptContent)
            }
        }
    }
    
    override func serviceExtensionTimeWillExpire() {
        if let contentHandler = self.contentHandler, let bestAttemptContent = self.bestAttemptContent {
            if MDMNotification.isMDMNotification(bestAttemptContent.userInfo) {
                super.serviceExtensionTimeWillExpire()
            } else {
                // Your code here
                contentHandler(bestAttemptContent)
            }
        }
    }

}
